zproject/backends.py

```diff
@@ -1018,6 +1021,38 @@ def sync_groups_from_ldap(self, user_profile: UserProfile, ldap_user: _LDAPUser)
                     groups_for_membership_deletion, [user_profile.id], acting_user=None
                 )
 
+            if settings.LDAP_GROUP_DESCRIPTION_ATTR is None:
+                # Syncing of group descriptions is not configured.
+                return
+
+            ldap_group_descriptions_dict = ldap_user._get_group_descriptions()
+            zulip_group_descriptions_dict = {
+                group.name: group.description
+                for group in NamedUserGroup.objects.filter(
+                    realm=user_profile.realm, name__in=configured_ldap_group_names_for_sync
+                )
+            }
+
+            for group_name, ldap_description in ldap_group_descriptions_dict.items():
+                if group_name in intended_group_name_set_for_user:
+                    zulip_description = zulip_group_descriptions_dict.get(group_name)
+                    if ldap_description is not None and zulip_description != ldap_description:
+                        # Sanity assert: make sure we didn't get some strange data from LDAP.
+                        assert isinstance(ldap_description, str)
+
+                        user_group = NamedUserGroup.objects.get(
+                            name=group_name, realm=user_profile.realm
+                        )
+                        ldap_logger.debug(
+                            "Updating group description for %s from %s to %s",
+                            group_name,
+                            zulip_description,
+                            ldap_description,
+                        )
+                        do_update_user_group_description(
+                            user_group, ldap_description, acting_user=None
+                        )
```

Comment:
It looks like this does this loop for each user being synced, which is a lot of useless duplicate work. I think we need this to be a new top-level function, called directly from `sync_ldap_user_data` outside the loop.

Reply:
@timabbott This description-sync indeed gets run per user, but I don't think it's that much duplication? But the worst thing that gets run like is the group fetch
```
            zulip_group_descriptions_dict = {
                group.name: group.description
                for group in NamedUserGroup.objects.filter(
                    realm=user_profile.realm, name__in=configured_ldap_group_names_for_sync
                )
            }
```

which is probably not a big? The actual updates don't get duplicated because the loop checks the Zulip group description need to be updated. So any group which needs updating, will get updated once and that's all.

I agree that the natural way for this functionality to work would be outside of the users loop, as a separate sync ldap task which just:
1. Fetches all groups in LDAP with the description values.
2. Fetches all ldap-syncable Zulip groups
3. Does the appropriate sync.

But I don't know if `django-auth-ldap` provides an api for doing (1). Maybe i missed something obvious, but didn't find. Because it's focused on being a user auth library, its view of operating on groups is  "groups this specific user belongs to". So we actually have to work in a per-user loop because we can only fetch groups per-user.

Reply:
I think if you've got 10K users and 25 groups, it seems bad to do fetch the 25 groups 10K times. Probably we can just have a `already_synced_groups` set that keeps track of which ones we've processed in this run?

Reply:
@timabbott  We don't have a very good way to do that, this is called inside of `ZulipLDAPUserPopulator.get_or_build_user`, which is called inside `django-auth-ldap` - so we don't have the control to pass state like this from iteration to iteration. We could probably work out something very hacky  like tacking on the `.already_synced_groups` attribute on the `ZulipLDAPUser` object being synced  - which could then be read from inside `get_or_build_user`. 

So that sounds ugly enough that I left it out for now. Just made the query a bit better instead:
```
            zulip_group_descriptions_dict = {
                group.name: group.description
                for group in NamedUserGroup.objects.filter(
                    # We only need to fetch the groups for which the user is a membr
                    # of the corresponding LDAP group. We don't have ldap description data for other
                    # groups, so they won't be subject to syncing anyway.
                    realm=user_profile.realm,
                    name__in=ldap_group_name_set_for_user,
                ).only("name", "description")
            }
```